pymacaroons/serializers/binary_serializer.py

```
@@ -113,3 +236,103 @@ def _depacketize(self, packet):
+        if packet.field_type == _EOS:
+            return rest, packets
+        if packet.field_type <= prev_field_type:
+            raise MacaroonDeserializationException('fields out of order')
+        packets.append(packet)
+        prev_field_type = packet.field_type
+        data = rest
+
+
+def _parse_packet_v2(data):
+    ''' Parses a V2 data packet at the start of the given data.
+
+    The format of a packet is as follows:
+
+    field_type(varint) payload_len(varint) data[payload_len bytes]
+
+    apart from EOS which has no payload_en or data (it's a single zero byte).
+
+    :param data:
+    :return: rest of data, PacketV2
+    '''
+    from pymacaroons.exceptions import MacaroonDeserializationException
+
+    ft, n = _decode_uvarint(data)
+    data = data[n:]
+    if ft == _EOS:
+        return data, PacketV2(ft, None)
+    payload_len, n = _decode_uvarint(data)
+    data = data[n:]
+    if payload_len > len(data):
+        raise MacaroonDeserializationException(
+            'field data extends past end of buffer')
+    return data[payload_len:], PacketV2(ft, data[0:payload_len])
+
+
+def _encode_uvarint(data, n):
+    ''' Encodes integer into variable-length format into data.'''
+    if n < 0:
+        raise ValueError('only support positive integer')
+    while True:
+        this_byte = n & 127
+        n >>= 7
+        if n == 0:
+            data.append(this_byte)
+            break
+        data.append(this_byte | 128)
+
+
+def _decode_uvarint(data):
+    ''' Decode a variable -length integer.
+
+    Reads a sequence of unsigned integer byte and decodes them into an integer
+    in variable-length format and returns it and the length read.
+    '''
+    n = 0
+    shift = 0
+    length = 0
+    for b in data:
+        if not isinstance(b, int):
+            b = six.byte2int(b)
+        if b < 0x80:
+            return n | b << shift, length + 1
+        n |= (b & 0x7f) << shift
+        shift += 7
+        length += 1
+    return n, length
```

If we've got here, the varint is invalid, so we should raise an exception, I think.